src/StreamingDataFrames/streamingdataframes/state/rocksdb/stores.py

```diff
@@ -0,0 +1,432 @@
+                    options=self._options.to_options(),
+                    access_type=rocksdict.AccessType.read_write(),
+                )
+                logger.info(
+                    f'Successfully opened db partition on "{self._path}"',
+                )
+                return db
+            except Exception as exc:
+                is_locked = str(exc).lower().startswith("io error: lock")
+                if not is_locked:
+                    raise
+
+                if self._open_max_retries <= 0 or attempt >= self._open_max_retries:
+                    raise
+
+                logger.warning(
+                    f"Failed to open db partition, cannot acquire a lock. "
+                    f"Retrying in {self._open_retry_backoff}sec."
+                )
+
+                attempt += 1
+                time.sleep(self._open_retry_backoff)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+
+class TransactionStore:
+    """
+    A transaction-based store to perform simple key-value operations like
+    "get", "set", "delete" and "exists".
+    It is supposed to be used by stateful operators in Streaming DataFrames.
+
+
+    Serialization
+    *************
+    `TransactionStore` automatically serializes keys and values to JSON.
+
+    Prefixing
+    *********
+    `TransactionStore` allows to set prefixes for the keys in the given code block
+    using :meth:`with_prefix()` context manager.
+    Normally, Streaming DataFrames will set prefixes order to scope
+    the store operations
+
+    Transactional properties
+    ************************
+    `TransactionStore` uses a combination of in-memory update cache
+    and RocksDB's WriteBatch in order to accumulate all the state mutations
+    in a single batch, flush them atomically, and allow the updates be visible
+    within the transaction before it's flushed (aka "read-your-own-writes" problem).
+
+    A single transaction can span across multiple incoming messages,
+
+    If any mutation fails during the transaction
+    (e.g. we failed to write the updates to the RocksDB), the whole transaction is now
+    invalid and cannot be used anymore.
+    In this case, a new `TransactionStore` should be created.
+
+    The `TransactionStore` also cannot be used after it's flushed, and a new
+    instance of `TransactionStore` should be used instead.
+
+    :param storage: instance of `StateStorage` to be used for accessing
+        the underlying RocksDB
+
+    """
+
+    __slots__ = (
+        "_storage",
+        "_update_cache",
+        "_batch",
+        "_prefix",
+        "_failed",
+        "_completed",
+        "_dumps",
+        "_loads",
+    )
+
+    def __init__(
+        self,
+        storage: RocksDBStorage,
+        dumps: Optional[DumpsFunc] = None,
+        loads: Optional[LoadsFunc] = None,
+    ):
+        self._storage = storage
+        self._update_cache = {}
```

Comment:
There might be some misunderstanding about the purpose of the update cache here.

The only purpose of the update cache is to enable a reading of the same values that have been previously updated in the same transaction (aka "reading-your-own-writes" problem).
`rocksdict.WriteBatch` doesn't allow you to check the pending updates before they're flushed to the db.

Optimization of reads is not a goal here.

Example:
```
store = TransactionStore()
store.set('key', 1)  # Set a "key" to "1"
store.get('key')  # Must return "1". 

# Without update cache, it will return whatever value is set in the DB right now,
# not the updated one
```

Reply:
Sorry, I totally understand the confusion here...was worried about my comment implying this. 

Though I was in fact talking about optimization of reads as a general idea in my comment, I do understand that's not the purpose of this cache here (which is to just read your own writes until they are actually committed to the DB). 

My original comment was meant as a more general consideration/thought around caching in general. I should have put it as a general comment, sorry!